docs/architecture/adr-062-collections-state-layer.md

```diff
@@ -0,0 +1,117 @@
+
+### Current Solution: ORM
+
+The current SDK proposed solution to this problem is [ORM](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-055-orm.md).
+Whilst ORM offers a lot of good functionality aimed at solving these specific problems, it has some downsides:
+- It requires migrations.
+- It uses the newest protobuf golang API, whilst the SDK still mainly uses gogoproto. 
+- Integrating ORM into a module would require the developer to deal with two different golang frameworks (golang protobuf + gogoproto) representing the same API objects.
+- It has a high learning curve, even for simple storage layers as it requires developers to have knowledge around protobuf options, custom cosmos-sdk storage extensions, and tooling download. Then after this they still need to learn the code-generated API.
+
+### CosmWasm Solution: cw-storage-plus
+
+The collections API takes inspiration from [cw-storage-plus](https://docs.cosmwasm.com/docs/1.0/smart-contracts/state/cw-plus/),
+which has demonstrated to be a powerful tool for dealing with storage in CosmWasm contracts.
+It's simple, does not require extra tooling, it makes it easy to deal with complex storage structures (indexes, snapshot, etc).
+The API is straightforward and explicit.
+
+## Decision
+
+We propose to port the `collections` API, whose implementation lives in [NibiruChain/collections](https://github.com/NibiruChain/collections) to cosmos-sdk.
+
+Collections implements four different storage handlers types:
+
+- `Map`: which deals with simple `key=>object` mappings.
+- `KeySet`: which acts as a `Set` and only retains keys and no object (usecase: allow-lists).
+- `Item`: which always contains only one object (usecase: Params)
+- `Sequence`: which implements a simple always increasing number (usecase: Nonces)
+- `IndexedMap`: builds on top of `Map` and `KeySet` and allows to create relationships with `Objects` and `Objects` secondary keys.
+
+All the collection APIs build on top of the simple `Map` type.
+
+Collections is fully generic, meaning that anything can be used as `Key` and `Value`. It can be a protobuf object or not.
+
+Collections types, in fact, delegate the duty of serialisation of keys and values to a secondary collections API component called `ValueEncoders` and `KeyEncoders`.
+
+`ValueEncoders` take care of converting a value to bytes (relevant only for `Map`). And offers a plug and play layer which allows us to change how we encode objects, 
+which is relevant for swapping serialisation frameworks and enhancing performance.
+`Collections` already comes in with default `ValueEncoders`, specifically for: protobuf objects, special SDK types (sdk.Int, sdk.Dec).
+
+`KeyEncoders` take care of converting keys to bytes, `collections` already comes in with some default `KeyEncoders` for some privimite golang types
```

Comment:
One downside in my opinion of KeyEncoders being an open interface is that the number of encoding formats is unbounded.

For each new custom KeyEncoder that a chain decides to create, the same code would need to be rewritten in JS, to be used for light client proofs.

In the proto ORM, the number of encoders was standardized, and all info to understand key format was in proto files (queryable by clients), so it was "write once use forever" for clients. In this ADR, afaiu, reflection only returns a string of the key encoder type, so the client would need a custom registry of key encoder type to implementation, and possibly write new implementations for custom chains.

Hopefully in practice, the community would converge to a set of 10-12 key encoders. (same applies for value encoders)

Reply:
Yes, that is a downside. Hopefully we can still have a standardized set of key + value encoders to simplify client integration.

Reply:
Yes this is an unfortunate downside, but a desired property in order to incentivise developers to migrate in a backwards compatible way with respect to their state.

In the future nothing forbids us from adding a `doNotImplement` method on `KeyCodec`, I even wonder if this could be done gradually, type by type; example: starting from uint64, string...